internal/gatewayapi/translator.go

```diff
@@ -901,108 +1004,35 @@ func (t *Translator) ProcessHTTPRoutes(httpRoutes []*v1beta1.HTTPRoute, gateways
-						)
-						continue
-					}
 
-					if backendRef.Kind != nil && *backendRef.Kind != KindService {
-						parentRef.SetCondition(
-							v1beta1.RouteConditionResolvedRefs,
-							metav1.ConditionFalse,
-							v1beta1.RouteReasonInvalidKind,
-							"Kind is invalid, only Service is supported",
-						)
-						continue
-					}
+					destination := buildRuleRouteDest(backendRef, parentRef, httpRoute, resources)
 
-					if backendRef.Namespace != nil && string(*backendRef.Namespace) != "" && string(*backendRef.Namespace) != httpRoute.Namespace {
-						if !isValidCrossNamespaceRef(
-							crossNamespaceFrom{
-								group:     v1beta1.GroupName,
-								kind:      KindHTTPRoute,
-								namespace: httpRoute.Namespace,
-							},
-							crossNamespaceTo{
-								group:     "",
-								kind:      KindService,
-								namespace: string(*backendRef.Namespace),
-								name:      string(backendRef.Name),
-							},
-							resources.ReferenceGrants,
-						) {
-							parentRef.SetCondition(
-								v1beta1.RouteConditionResolvedRefs,
-								metav1.ConditionFalse,
-								v1beta1.RouteReasonRefNotPermitted,
-								fmt.Sprintf("Backend ref to service %s/%s not permitted by any ReferenceGrant", *backendRef.Namespace, backendRef.Name),
-							)
-							continue
+					for _, route := range ruleRoutes {
+						// If the route already has a direct response or redirect configured, then it was from a filter so skip
+						// processing any destinations for this route.
+						if route.DirectResponse == nil && route.Redirect == nil {
+							if destination != nil {
+								route.Destinations = append(route.Destinations, destination)
+								continue
+							} else {
+								route.InvalidBackends += 1
+							}
 						}
 					}
 
-					if backendRef.Port == nil {
-						parentRef.SetCondition(
-							v1beta1.RouteConditionResolvedRefs,
-							metav1.ConditionFalse,
-							"PortNotSpecified",
-							"A valid port number corresponding to a port on the Service must be specified",
-						)
-						continue
-					}
-
-					service := resources.GetService(NamespaceDerefOr(backendRef.Namespace, httpRoute.Namespace), string(backendRef.Name))
-					if service == nil {
-						parentRef.SetCondition(
-							v1beta1.RouteConditionResolvedRefs,
-							metav1.ConditionFalse,
-							v1beta1.RouteReasonBackendNotFound,
-							fmt.Sprintf("Service %s/%s not found", NamespaceDerefOr(backendRef.Namespace, httpRoute.Namespace), string(backendRef.Name)),
-						)
-						continue
-					}
+				}
 
-					var portFound bool
-					for _, port := range service.Spec.Ports {
-						if port.Port == int32(*backendRef.Port) {
-							portFound = true
-							break
+				// If the route has no valid backends then just use a direct response and don't fuss with weighted responses
+				for _, ruleRoute := range ruleRoutes {
+					if ruleRoute.InvalidBackends > 0 && len(ruleRoute.Destinations) == 0 {
```

Comment:
why not set the DirectResponse status in L1018 ?

Reply:
Because at that point we've not figured out if the backends are valid or not yet so we can't determine if the `DirectResponse` will be used at all.